src/pages/index.js

```diff
@@ -93,34 +117,120 @@ function setFormInfo(nameSelector, detailsSelector) {
 }
 
 function createCard(data) {
-  const cardElement = new Card({ data, handleImageClick }, "#cards-template");
+  const cardElement = new Card(
+    { data, handleImageClick },
+    "#cards-template",
+    deleteCardConfirm,
+    handleCardLike
+  );
   return cardElement.generateCard();
 }
 
+function deleteCardConfirm(id, cardElement) {
+  deleteCardConfirmModal.open();
+  deleteCardConfirmModal.setSubmitAction(() => {
+    handleDeleteCard(id, cardElement);
+  });
+}
+
 function handleImageClick(imgData) {
   previewModal.open(imgData);
 }
 
+function handleCardLike(id, likeFunction, unlikeFunction, button, card) {
+  if (!card._isLiked) {
+    pageInfo.cardLike(id).then((res) => {
+      likeFunction(button, card);
+    });
+  } else {
+    pageInfo.cardUnlike(id).then((res) => {
+      unlikeFunction(button, card);
+    });
+  }
+}
+
+function handleDeleteCard(id, cardElement) {
+  pageInfo
+    .deleteCard(id)
+    .then((res) => {
+      cardElement.remove();
+      cardElement = null;
+      deleteCardConfirmModal.close();
+    })
+    .catch((err) => console.log(err))
+    .finally(() => {
+      deleteCardConfirmModal.removeProcessingMessage("Yes");
+    });
+}
+
 function handleProfileFormSubmit(evt) {
   evt.preventDefault();
-  const { name, description } = profileEdit.formValues;
-  updateUserInfo(profileEdit.formValues);
-
-  profileEdit.close();
-  formValidators["profile-edit-form"].resetValidation();
+  const submittedUserInfo = profileEdit.formValues; //REVIEWER COMMENT HERE FOR MAKING SUBMIT FUNCTION
+  const { name, description: about } = submittedUserInfo;
+  pageInfo
+    .changeUserInfo({ name, about })
+    .then((response) => {
+      const name = response.name;
+      const description = response.about;
+      updateUserInfo({ name, description });
+      return response;
+    })
+    .then((res) => {
+      console.log(res);
+      profileEdit.close();
+    })
+    .then((res) => {
+      formValidators["profile-edit-form"].resetValidation();
+      profileEdit.resetForm();
+    })
+    .catch((err) => console.log(err))
+    .finally(() => {
+      profileEdit.removeProcessingMessage("Save");
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```

Reply:
Hello! I tried to implement this and I'm having some difficulty with it. I have my current iteration of the code commented out on my latest commit. The biggest issue is that it looks like I'm unable to access "this" when I call the api request function. If you think the code being consolidated in this way is worth the effort, or you have any tips for how I might be able to better maneuver the code, let me know! Thanks for your help! @gennady-bars